### ISSUE TITLE
Generate POM for non-Maven Clojure projects

### ISSUE THREAD

**Issue by pmonks** [jlesquembre/clojars-publish-action#1]
For non-Maven Clojure projects, having a static `pom.xml` file checked into the repository is an anti-pattern (it violates the [DRY Principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)).

For Leiningen and tools.deps projects specifically (which covers a large majority of Clojure projects), tooling is available to generate a transient `pom.xml` file on demand that can then be transmitted to Clojars.  For Leiningen, the command is `lein pom`, while for tools.deps (with the Clojure CLI tools installed) the command is `clj -Spom`.

It would be preferable if this GitHub Workflow Action was to use these tools to generate a transient `pom.xml` file as part of the Clojars deployment process for Leiningen and tools.deps based projects, rather than relying on the existence of a static one in the project's committed assets.

**Comment by jlesquembre** [comment]
Thanks for opening an issue!

I only support tools.deps projects in this action. I don't have plans to add support for Leiningen, but if you want to work on it, I'll be happy accepting a PR and helping you.

I don't think that having a `pom.xml` file in the repo is an anti-pattern. Notice that I'm using the `pom.xml` as a template, and filling the information about dependencies with `clj -Spom`. The information about the version is taken from git. I'm updating the README to try to make it clearer.

If you just run `clj -Spom`, the generated `pom.xml` is going to miss some data, like project name, description, etc. 

Here you can see some examples of other people adding a `pom.xml` to their projects:
https://github.com/stuarthalloway/clojure-from-java
https://oli.me.uk/clojure-projects-from-scratch/#_publishing_to_clojars
https://github.com/seancorfield/depstar

**Comment by pmonks** [comment]
If you look more closely at projects that use this approach, you'll see that there is indeed a violation of the DRY Principle. [seancorfield/depstar](https://github.com/seancorfield/depstar), to pick one of your examples, has these duplicates:

* [Dependency in `deps.edn`](https://github.com/seancorfield/depstar/blob/develop/deps.edn#L2)
* [Duplicate dependency in `pom.xml`](https://github.com/seancorfield/depstar/blob/develop/pom.xml#L31-L35)

Although `depstar` only has one dependency (and hence this looks like a trivial problem), it's easy to imagine the significant maintenance headaches (not to mention unexpected runtime behaviour when these files end up unsynchronised) that would be incurred for a project that has a more typical number of dependencies.

To pick another example, the small hobby project I'm currently working on (a relatively small thing; < 500 LOC) currently has 19 direct dependencies and 62 additional transitive dependencies, and even at that small scale manually maintaining those dependencies in duplicate would be unmanageable.

And yes it's somewhat problematic that `deps.edn` doesn't (yet) have a standardised structure for project metadata beyond the things necessary for classpath construction (tools.deps' current laser focus).  That said, the `deps.edn` file is in EDN format, so there's no reason your workflow action couldn't expect or require additional elements in `deps.edn` (e.g. the project's Maven-style GAV) that tools.deps remains oblivious to.  I'd argue that embracing and extending `deps.edn` as **the** source of project metadata is better than expecting developers to manually duplicate information between multiple files.  Of course the downside is that this also means that the workflow action would have to generate a more comprehensive `pom.xml` from `deps.edn`, since the built-in `clj -Spom` isn't going to be sensitive to the "bonus" elements in `deps.edn`.

**Comment by jlesquembre** [comment]
Most probably depstar's author is using `clj -Spom`to update the dependencies in `pom.xml`, I don't think that he is managing the file manually. 

I don't see duplicated information between multiple files, but the information is split in 2 files. `deps.edn` has the information about dependencies, while `pom.xml` has only the metadata for the project, see the example in the readme:
https://github.com/jlesquembre/clojars-publish-action/blob/1c159e1d261743f9c1c8dc89b4ef358f493d5f5e/README.md#L9-L43

I agree that it would be nice to have all the information in `deps.edn`, but to do that I'll need to write extra code for parsing it and update the `pom.xml`. I prefer to leverage the existing tools, maven and clj CLI. The only drawback is that you need to create a `pom.xml` with your project's meta-data, but for me it is acceptable. Keep in mind that you need to do that only once, you can just copy/paste it from the readme, and you don't need to update the `pom.xml` ever again. I prefer that to the alternative of adding an extra dependency to generate a `pom.xml` from a non-standard `deps.edn` file

I created a dummy project to test the action:
https://github.com/jlesquembre/dummy-clj
Maybe you find useful a real example

**Comment by pmonks** [comment]
I didn't realise `clj -Spom` will merge in dependencies etc. to an existing "partial" `pom.xml`, if one exists.  #TIL  That is indeed a much better situation (no duplication needed).

One thing in the updated readme that caught my eye though:

> Also, `version` and `smc/tag` values don't matter, since they will be replaced with the data from git (tag or branch name).

Relying on git branch and/or tag names for version information seems problematic, as it assumes a lot about how git is used by project maintainers.  In particular I'm not sure how it could work with SNAPSHOT releases, without renaming the `main` (aka `master`) branch every single time a new release is cut (or ending up with versions such as `-SNAPSHOT`, which is also problematic in other ways).

For example, I might have a project where I want all commits / merges to `main` to be deployed as a SNAPSHOT to Clojars, using the next release's version number as a prefix (e.g. if the current released version is `1.0.0`, the snapshot versions should be either `1.0.1-SNAPSHOT`, `1.1.0-SNAPSHOT`, or `2.0.0-SNAPSHOT`, at my discretion).

How would I achieve this with an approach based on branch or tag names, without being forced to rename my `main` branch every time I make a release?

**Comment by jlesquembre** [comment]
Currently is not possible to create snapshots with the next release version number, but it creates snapshots with the branch name.

Maybe is easier to understand with an example. Every time you create a new tag, e.g. `1.0.0` a new version is deployed with that number. Snapshots are created when you push to a branch. Let's say that you have 2 branches, `main` and `dev`, if you push to `main`, a new release, `main-SNAPSHOT` is deployed, while if you push to `dev`, `dev-SNAPSHOT` is deployed.
I know that is not standard in the maven ecosystem to have that version names, but I think it makes more sense for several reasons:
- It makes clearer what is the version in the snapshot. If you are working on branch `foo` and `bar` for the next release, what version will you get from `1.0.1-SNAPSHOT`? `foo` or `bar`? You don't know, it depends on the latest push 
- You can't be sure about your next version number. Maybe you are thinking about releasing `1.0.1`, but at the end some breaking changes are needed and you need to release version `2.0.0`
- I find useless to have that many snapshot versions, `1.0.1-SNAPSHOT`, `1.0.2-SNAPSHOT`, `1.0.3-SNAPSHOT`... After you release the next version, who cares about previous snapshots? If I'm on version `1.1.0`, who needs to install `1.0.X-SNAPSHOT`?


I'll try to add a better explanation in the readme tomorrow.

**Comment by pmonks** [comment]
Yes this is as I understood the situation, although more & clearer documentation is always a good thing.

The concern I have is that this is a highly unusual approach to snapshot version numbering that will introduce problems for downstream consumers, since:
1. it's not compatible with [Maven's version comparison logic](https://github.com/apache/maven/blob/master/maven-artifact/src/main/java/org/apache/maven/artifact/versioning/ComparableVersion.java), which is the same logic Leiningen uses (and I'd be surprised if tools.deps deviated from it, given the difficulties that would introduce when it has to access Maven repository-hosted dependencies)
2. it doesn't conform to generally accepted versioning best practices (i.e. [SemVer](https://semver.org/)).  More specific to the Clojure ecosystem, see also the 3rd recommendation in the [Leiningen FAQ](https://github.com/technomancy/leiningen/blob/master/doc/FAQ.md).

Point 1 isn't a hypothetical issue either - it's easy to start to get a sense for the implications in a REPL:
```
Script started on Wed Jun 24 14:57:25 2020
bash-3.2$ clj -Sdeps '{:deps {org.apache.maven/maven-artifact #:mvn {:version "3.6.3"}}}'
Clojure 1.10.1
user=> (import org.apache.maven.artifact.versioning.ComparableVersion)
org.apache.maven.artifact.versioning.ComparableVersion
user=> (def ver-main-snapshot (ComparableVersion. "main-SNAPSHOT"))
#'user/ver-main-snapshot
user=> (def ver-010 (ComparableVersion. "0.1.0"))
#'user/ver-010
user=> (def ver-100 (ComparableVersion. "1.0.0"))
#'user/ver-100
user=> (def ver-253 (ComparableVersion. "2.5.3"))
#'user/ver-253
user=> (def ver-9900 (ComparableVersion. "99.0.0"))
#'user/ver-9900
user=> (def ver-9900-snapshot (ComparableVersion. "99.0.0-SNAPSHOT"))
#'user/ver-9900-snapshot
user=> (.compareTo ver-main-snapshot ver-010)
-1
user=> (.compareTo ver-main-snapshot ver-100)
-1
user=> (.compareTo ver-main-snapshot ver-253)
-1
user=> (.compareTo ver-main-snapshot ver-9900)
-1
user=> (.compareTo ver-main-snapshot ver-9900-snapshot)
-1
user=> ^D
bash-3.2$ exit
Script done on Wed Jun 24 14:58:38 2020
```

**Comment by jlesquembre** [comment]
I don't think that it will introduce problems for downstream consumers. Most of the time, you want to use the release versions. I want to use snapshot versions only in very special situations. I prefer to avoid them, since there is no guarantees that version `1.0.0-SNAPSHOT` will be the same today and tomorrow.
 
Also, with clojure libraries, if I want to use an unreleased version, I prefer to use the git coordinate. That way I can be sure about the version I get, and I don't need to wait until the author published a snapshot. 
 
Going back to your concerns, why do you want to compare snapshot versions against release versions? I cannot think of an use case for that, do you have one?

If really needed, I think I could add an option to follow the common approach for snapshots. Probably the best approach would be to find the latest tag and increment the minor version, if your latest tag is `1.0.0`, your snapshot version would be `1.0.1-SNAPSHOT`. For version numbers, I want to use the information already on git, I never liked to maintain a file with the version number in sync with git tags. From my point of view, git must be the only source of true. 

One last note, even if most people is using SemVer, there are problems with it. There is a very interesting talk from Rich Hickey about it: https://www.youtube.com/watch?v=oyLBGkS5ICk

**Comment by pmonks** [comment]
> Most of the time, you want to use the release versions. I want to use snapshot versions only in very special situations. I prefer to avoid them, since there is no guarantees that version `1.0.0-SNAPSHOT` will be the same today and tomorrow.

Of course, but there are times when I do indeed want to consume a SNAPSHOT - for example when I want to test my app against an upcoming (but not yet released) new version of an upstream dependency.  In fact I happen to be doing that right now with one of the dependencies in my hobby project (and helping the author of that library test & polish the code ahead of their next release - a win/win).

There are also cases where library authors are lackadaisical about releases, and all you've got to work with is a deployed SNAPSHOT (if that).  Another of my hobby project's dependencies falls into this camp, and until this week it wasn't accessible via tools.deps' git/SHA coordinates at all because it lacked a `deps.edn` file (it has one now, but only because I submitted a PR for it - it would have been even better if the author had issued a new release, but I get the sense that they don't have the time and/or interest for that, and so my only actionable alternative was to contribute a `deps.edn` file, which solves my problem but doesn't help Maven / Leiningen users who consume the library).

> Also, with clojure libraries, if I want to use an unreleased version, I prefer to use the git coordinate. That way I can be sure about the version I get, and I don't need to wait until the author published a snapshot.

Yeah git/SHA coordinates are great, but they aren't always applicable for a couple of reasons:
1. Not all Clojure projects have a `deps.edn` file in their repository, and without one tools.deps can't access them via git/SHA coordinates (the second example I mentioned above had this problem until earlier this week).
2. Approximately zero non-Clojure projects will ever get a `deps.edn` file, and even if they did, it's unlikely tools.deps would be able to build that code locally anyway.
3. This mechanism is specific to tools.deps and doesn't exist* in some other popular Java ecosystem build tools (Ant+Ivy, Maven, Leiningen, etc.).

*I'm aware of (and use) [jitpack.io](https://jitpack.io/), but it's not a standard part of those tools and has other issues

> Going back to your concerns, why do you want to compare snapshot versions against release versions? I cannot think of an use case for that, do you have one?

Because version numbers, including SNAPSHOT versions, have a well defined conceptual order.  Here's an example (from lowest/oldest to highest/newest):
* `0.0.1-SNAPSHOT`
* `1.0.0-RC1`
* `1.0.0-RC2`
* `1.0.0`
* `1.0.1-SNAPSHOT`
* `1.0.1`
* `1.1.0`
* `2.0.0-RC1`
* `2.0.0-SNAPSHOT`
* `2.0.0`

And we can demonstrate that `org.apache.maven.artifact.versioning.ComparableVersion` (the underlying implementation for Maven and Leiningen, at least) faithfully implements this ordering:

```
$ clj -Sdeps '{:deps {org.apache.maven/maven-artifact #:mvn {:version "3.6.3"}}}'
Clojure 1.10.1
user=> (import org.apache.maven.artifact.versioning.ComparableVersion)
org.apache.maven.artifact.versioning.ComparableVersion
user=> (def versions [(ComparableVersion. "1.0.1") (ComparableVersion. "0.0.1-SNAPSHOT") (ComparableVersion. "1.0.0-RC2") (ComparableVersion. "2.0.0-RC1") (ComparableVersion. "1.1.0") (ComparableVersion. "1.0.0") (ComparableVersion. "1.0.1-SNAPSHOT") (ComparableVersion. "1.0.0-RC1") (ComparableVersion. "2.0.0-SNAPSHOT") (ComparableVersion. "2.0.0")])  ; Random ordering to prove that sorting works as described
#'user/versions
user=> (map str (sort versions))
("0.0.1-SNAPSHOT" "1.0.0-RC1" "1.0.0-RC2" "1.0.0" "1.0.1-SNAPSHOT" "1.0.1" "1.1.0" "2.0.0-RC1" "2.0.0-SNAPSHOT" "2.0.0")
```

In contrast, a non-standard version number such as `main-SNAPSHOT` does not fit in this ordering, and indeed the Maven implementation falls back by making it the lowest/oldest version number in all cases (which would not be correct in the scheme you proposed above):

```
user=> (def versions [(ComparableVersion. "1.0.1") (ComparableVersion. "0.0.1-SNAPSHOT") (ComparableVersion. "1.0.0-RC2") (ComparableVersion. "2.0.0-RC1") (ComparableVersion. "1.1.0") (ComparableVersion. "1.0.0") (ComparableVersion. "1.0.1-SNAPSHOT") (ComparableVersion. "1.0.0-RC1") (ComparableVersion. "main-SNAPSHOT") (ComparableVersion. "2.0.0-SNAPSHOT") (ComparableVersion. "2.0.0")])  ; Random ordering to prove that sorting works as described
#'user/versions
user=> (map str (sort versions))
("main-SNAPSHOT" "0.0.1-SNAPSHOT" "1.0.0-RC1" "1.0.0-RC2" "1.0.0" "1.0.1-SNAPSHOT" "1.0.1" "1.1.0" "2.0.0-RC1" "2.0.0-SNAPSHOT" "2.0.0")
```

There are a variety of use cases that depend on total ordering of version numbers, including:
1. dependency version conflict detection/resolution (something tools.deps doesn't support currently, but will hopefully eventually offer, similar to how Maven, Leiningen etc. do)
2. reporting out of date dependencies (something tools.deps doesn't do, but for which there's a [plugin](https://github.com/Olical/depot))
3. use of version number ranges for upstream dependencies (something tools.deps doesn't support, but other build tools do)
4. ...and more...

But zooming out for a moment, while it's valuable to make a good argument for why the Java ecosystem's de facto standard for representing SNAPSHOT versions is "wrong", I personally think that there are more effective ways to make that case than by creating a tool that's deliberately incompatible with it.  All that does is reduce the value of the tool to its potential users (whether they agree with your argument or not).

> If really needed, I think I could add an option to follow the common approach for snapshots. Probably the best approach would be to find the latest tag and increment the minor version, if your latest tag is `1.0.0`, your snapshot version would be `1.0.1-SNAPSHOT`. For version numbers, I want to use the information already on git, I never liked to maintain a file with the version number in sync with git tags. From my point of view, git must be the only source of true.

This may be useful for some users of the tool, and it may even be something I'd end up using myself once I have time to see how well it integrates with my workflow.  But in the meantime, I would prefer to have direct control over the version numbers of my deployed artifacts, and I _think_ that the simplest way to do that would be to provide an option in the tool whereby it simply plucks the version number verbatim from the `<version>` tag in the project's partial `pom.xml`, and doesn't try to automagically infer version numbers from state that is harder to control i.e. tag and (especially) branch names.

> One last note, even if most people is using SemVer, there are problems with it. There is a very interesting talk from Rich Hickey about it: [youtube.com/watch?v=oyLBGkS5ICk](https://www.youtube.com/watch?v=oyLBGkS5ICk)

Regardless, it is the only widely adopted version numbering standard I'm aware of, and is notably the de facto standard for the Java ecosystem (which Clojure is fundamentally coupled to, whether we like it or not).

**Comment by jlesquembre** [comment]
I didn't know about JitPack, looks interesting, thanks for sharing.
 
We have a different opinion here. Version number ordering for snapshots is not important for me. I'm not saying that the standard snapshot versioning in the java ecosystem is wrong, but it doesn't fit well in my workflow, that's why I don't use it.
 
The main reason I created this action was because I got tired of manually updating `pom.xml`, especially since that information was already on git. To me, re-using tags for version numbers is more DRY. Previously I used the maven-release-plugin, but I didn't like that workflow too much.
 
Keep in mind that I created this action just for me, and as far as I know, no one else uses it. While we don't agree, maybe I'll add an option to choose from where to pick the version number (git or pom.xml). I can't promise, but since it seems easy to do and you find value on it, maybe I'll do it, give me some time to think about it.

**Comment by pmonks** [comment]
Thanks - I do appreciate your willingness to consider other workflows!

**Comment by jlesquembre** [comment]
I thought a bit about it, and I'll add a new option (`use_git_ref` or something similar, suggestions welcome). It will be true by default, in which case the behavior will be the current one, but if you set it to false, it will just use the version number you have in your pom.xml. Let me know if you have something else in mind.

I'll do an update during the next days, I'm also doing some updates to the docker image used by the action

**Comment by pmonks** [comment]
That sounds great - thanks @jlesquembre!

**Comment by jlesquembre** [comment]
I just added a new option `USE_GIT_REF`. Set it to false and the version number will be taken from your pom.xml.
If you want to try it, use version `0.3`

I also rewrote the action to use clojure instead of bash. Now it would be easier to read some metadata from the `deps.edn` file.
I'm thinking about making the pom.xml file optional, and taking all the information directly from the `deps.edn` file.
